### PR TITLE
Check contract visibility when looking up contracts in sandbox

### LINKED ISSUE #10: check contract visibility when looking up contracts in sandbox
The scenario runner implements a key check making sure that contracts we look up are visible: https://github.com/digital-asset/daml/blob/2a932d9b229d178e9ae72ff1e33258650622e63b/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/types/Ledger.scala#L409 .

we need to implement the same check in the sandbox, but right now we don't, see https://github.com/digital-asset/daml/blob/2a932d9b229d178e9ae72ff1e33258650622e63b/ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/CommandExecutorImpl.scala#L31 .

to check, we need to check that the committer (which the sandbox knows) can see the contract requested (which the sandbox does not currently know).

to be able to do this, we need to start storing some additional data in `LedgerState`. specifically we need to keep track of which party can see contracts, and update this as we add transactions to the ledger state.

open question: do we need to take the ledger time into account at all in this visibility check? cc @oggy- @andreaslochbihler-da

### PR DESCRIPTION
**Pull request by stephencompall-DA**
Fixes #10.

### CODE REVIEW

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala [hunk 1]
```diff
@@ -200,30 +201,31 @@ private class PostgresLedgerDao(
...
+          storeContract(offset, Contract.fromActiveContract(cid, c))
+          this
+        }
+        //TODO: remove contract key
+        def removeContract(cid: AbsoluteContractId, keyO: Option[GlobalKey]) = {
+          archiveContract(offset, cid)
+          this
+        }
+        def implicitlyDisclose(global: Relation[AbsoluteContractId, Ref.Party]) =
+          this // TODO SC
```
Comment (francesco-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala]: we'll need an additional table here -- let's not do this in this pr.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveContracts.scala [hunk 2]
```diff
@@ -178,11 +193,19 @@ class ActiveContractsManager[ACS](
             }
         }
 
-    st.result
+    st.mapAcs(_ implicitlyDisclose globalImplicitDisclosure).result
   }
 
 }
 
+trait ActiveContractsSteps[+Self] { this: ActiveContractsSteps[Self] =>
```
Comment (francesco-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveContracts.scala]: this name is pretty bad now. suggestion:

```
ActiveContractsStep => ActiveContracts
ActiveContracts => ActiveContractsInMemory
```

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala [hunk 3]
```diff
@@ -233,6 +235,7 @@ private class PostgresLedgerDao(
         workflowId,
         transaction,
         mappedDisclosure,
+        Map.empty // TODO SC when do we enrich here?
```
Comment (francesco-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala]: you should pass the disclosure information in, even if for now we do nothing with it.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveContracts.scala [hunk 4]
```diff
@@ -178,11 +193,19 @@ class ActiveContractsManager[ACS](
...
+  def lookupContract(cid: AbsoluteContractId): Option[ActiveContract]
+  def keyExists(key: GlobalKey): Boolean
+  def addContract(cid: AbsoluteContractId, c: ActiveContract, keyO: Option[GlobalKey]): Self
+  def removeContract(cid: AbsoluteContractId, keyO: Option[GlobalKey]): Self
+  def implicitlyDisclose(global: Relation[AbsoluteContractId, Ref.Party]): Self
+}
+
 object ActiveContracts {
 
   case class ActiveContract(
```
Comment (francesco-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveContracts.scala]: please rename `witnesses` to `disclosedTo`, as discussed.

Comment (stephencompall-DA) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveContracts.scala]: fb861d97506edd7a1a809749b106a7c9dbd4c381

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveContracts.scala (hunk 2 above)
Comment (stephencompall-DA) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveContracts.scala]: e07c9390388e1db927546793f52dee963f422ffb

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala (hunk 1 above)
Comment (stephencompall-DA) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala]: 5dd7f557d90d82b19f68e09ad9d87dbb0b8082f6

**Comment by stephencompall-DA** [comment]
retest this please

**Comment by stephencompall-DA** [comment]
retest this please

**Pull request closed without merging**